include/pybind11/stl.h

```diff
@@ -35,6 +36,89 @@
 PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
 PYBIND11_NAMESPACE_BEGIN(detail)
 
+//
+// Begin: Equivalent of
+//        https://github.com/google/clif/blob/ae4eee1de07cdf115c0c9bf9fec9ff28efce6f6c/clif/python/runtime.cc#L388-L438
+/*
+The three `PyObjectTypeIsConvertibleTo*()` functions below are
+the result of converging the behaviors of pybind11 and PyCLIF
+(http://github.com/google/clif).
+
+Originally PyCLIF was extremely far on the permissive side of the spectrum,
+while pybind11 was very far on the strict side. Originally PyCLIF accepted any
+Python iterable as input for a C++ `vector`/`set`/`map` argument, as long as
+the elements were convertible. The obvious (in hindsight) problem was that
+any empty Python iterable could be passed to any of these C++ types, e.g. `{}`
+was accpeted for C++ `vector`/`set` arguments, or `[]` for C++ `map` arguments.
+
+The functions below strike a practical permissive-vs-strict compromise,
+informed by tens of thousands of use cases in the wild. A main objective is
+to prevent accidents and improve readability:
+
+- Python literals must match the C++ types.
+
+- For C++ `set`: The potentially reducing conversion from a Python sequence
+  (e.g. Python `list` or `tuple`) to a C++ `set` must be explicit, by going
+  through a Python `set`.
+
+- However, a Python `set` can still be passed to a C++ `vector`. The rationale
+  is that this conversion is not reducing. Implicit conversions of this kind
+  are also fairly commonly used, therefore enforcing explicit conversions
+  would have an unfavorable cost : benefit ratio; more sloppily speaking,
+  such an enforcement would be more annoying than helpful.
+*/
+
+inline bool PyObjectIsInstanceWithOneOfTpNames(PyObject *obj,
+                                               std::initializer_list<const char *> tp_names) {
+    if (PyType_Check(obj)) {
+        return false;
+    }
+    const char *obj_tp_name = Py_TYPE(obj)->tp_name;
+    for (const auto *tp_name : tp_names) {
+        if (std::strcmp(obj_tp_name, tp_name) == 0) {
+            return true;
+        }
+    }
+    return false;
+}
+
+inline bool PyObjectTypeIsConvertibleToStdVector(PyObject *obj) {
+    if (PySequence_Check(obj) != 0) {
+        return !PyUnicode_Check(obj) && !PyBytes_Check(obj);
+    }
+    return (PyGen_Check(obj) != 0) || (PyAnySet_Check(obj) != 0)
+           || PyObjectIsInstanceWithOneOfTpNames(
+               obj, {"dict_keys", "dict_values", "dict_items", "map", "zip"});
```

Comment:
This requires explicit subclasses, rather than any object that conforms to `collections.abc.Sequence`? (Likewise for `collections.abc.Mapping` and `collections.abc.Set` below) The Pythonic way to do this would be to ask if the object has the methods required by these protocols, rather than checking explicit inheritance.

Reply:
Specifically, from https://docs.python.org/3/library/collections.abc.html:

**`Sequence`**: `__getitem__`, `__len__`, `__contains__`, `__iter__`, `__reversed__`, `index`, and `count`

**`Set`**: 
`__contains__`, `__iter__`, `__len__`, `__le__`, `__lt__`, `__eq__`, `__ne__`, `__gt__`, `__ge__`, `__and__`, `__or__`, `__sub__`, `__xor__`, and `isdisjoint`

**`Mapping`**: `__getitem__`, `__iter__`, `__len__`, `__contains__`, `keys`, `items`, `values`, `get`, `__eq__`, and `__ne__`

Reply:
In addition to [what I wrote yesterday](https://github.com/pybind/pybind11/pull/4686#issuecomment-1813215017), there was actually another concern in the back of my mind: the current implementation is super fast, direct `strcmp` calls for 5 (`vector`) of 1 (`set`) very short string(s).

In contrast, `hasattr` is far more complex and probably orders of magnitude slower, although I wouldn't really worry about it in the usual bigger context (it'll not even be a blip on the radar in most cases). The code extra code complexity is the bigger worry.

So there is a code complexity and runtime overhead cost.

What is the benefit?

That brings me back to my "large-scale field experiment" experience hinted in the PR description. In a nutshell:

* Current pybind11 is super restrictive.

* Original PyCLIF was super permissive.

I had this huge pile of data from the wild contrasting the two. Based on that I could see very clearly that there is very little to gain from being more permissive than what this PR implements. IOW I don't think paying the code complexity and runtime overhead cost will help in any significant way.

Could you please let me know if you have a different opinion? — It will probably cost me a whole day worth of effort to implement and fully test (global testing) attribute inspection. We will also have to agree on what subsets of the attributes we actually want to test against. I expect that there will be some devils in the details.

Reply:
I think having a short path (check inheritance first) could alleviate much of the speed problem - the slower "correct" check could be done if this is not a subclass of dict. Also, this list is somewhat arbitrary - you probably included things (like `dict_keys`? `dict_items`?) based on usage (in PyCLIF), rather than making a comprehensive list of everything iterable in Python? Also, I believe checking the attributes on a class (you don't need to check them on the instance) is pretty fast (`typing.runtime_checkable` switched to [this in 3.12](https://docs.python.org/3/library/inspect.html#inspect.getattr_static) (since 3.2) which made it much faster).

From a usage standpoint, implementing the Protocol specified by these ABC's is supposed to make your class work wherever the standard library classes work, so it seems ideal if this could follow that. Not _required_ (and we could start with this if you really want to), but it seems like the correct thing to do.

> We will also have to agree on what subsets of the attributes we actually want to test against.

This is concretely specified in `collections.abc`. No arguments needed.

Reply:
> I think having a short path (check inheritance first) could alleviate much of the speed problem

I agree. Which basically means: this PR plus follow-on work to make the check more permissive.

Do you think it's reasonable then to merge this PR, and do the fallback attribute inspection in a separate PR? — To be completely honest, I wouldn't want to prioritize this, for the reasons explained before. It a significant amount of work, extra code complexity, and I don't think very many people will ever notice the difference in practice. But if someone else wants to invest the effort, I'd be on board.

> This is concretely specified in `collections.abc`. No arguments needed.

I see, sounds good to me, but from a purely practical viewpoint, assuming that most duck-typing only implements a subset of the attributes, insisting on the full set of attributes here probably means nobody will ever see a difference in practice, unless they are specifically looking to pass the checks here.